File: 01-Languages/CSharp/21-Solutions/solution-06.cs
// solution-06.cs - Concurrent Downloads Simulation
// See: ../20-Exercises/README.md#exercise-06--concurrent-downloads-simulation-advanced
//
// Run with:
//     dotnet run solution-06.cs

using System.Diagnostics;

var stopwatch = Stopwatch.StartNew();

// Starting all four Tasks here (not awaiting each one immediately) is what
// makes them run concurrently -- FetchAsync begins its Task.Delay the
// instant it's called, before this line even reaches WhenAll.
var t1 = FetchAsync("https://api.example.com/users", 300, shouldFail: false);
var t2 = FetchAsync("https://api.example.com/orders", 500, shouldFail: false);
var t3 = FetchAsync("https://api.example.com/broken", 200, shouldFail: true);
var t4 = FetchAsync("https://api.example.com/products", 400, shouldFail: false);
var tasks = new[] { t1, t2, t3, t4 };

try {
    await Task.WhenAll(tasks);
} catch (Exception) {
    // WhenAll rethrows only the FIRST faulted task's exception here --
    // deliberately ignored in this catch so the loop below can inspect
    // every task individually and report ALL failures, not just one.
}

stopwatch.Stop();
Console.WriteLine($"Elapsed: {stopwatch.ElapsedMilliseconds}ms (slowest single delay was 500ms -- proves concurrency, not a ~1400ms sum)");

Console.WriteLine("\n--- per-task results ---");
foreach (var task in tasks) {
    if (task.IsFaulted) {
        // Task.Exception wraps the real exception in an AggregateException;
        // unwrap via .InnerException (or .Flatten()) to get the actual
        // FetchFailedException and its message.
        Console.WriteLine($"  FAILED: {task.Exception!.InnerException!.Message}");
    } else {
        Console.WriteLine($"  OK: {task.Result}");
    }
}

async Task<string> FetchAsync(string url, int delayMs, bool shouldFail) {
    await Task.Delay(delayMs);
    if (shouldFail) {
        throw new FetchFailedException($"{url} -> request failed after {delayMs}ms");
    }
    return $"{url} -> 200 OK ({delayMs}ms)";
}

class FetchFailedException : Exception {
    public FetchFailedException(string message) : base(message) {}
}

File: 01-Languages/Dart/21-Solutions/solution-06.dart
// solution-06.dart - Exercise 06: Concurrent Futures.

class FetchFailedException implements Exception {
  final String url;
  FetchFailedException(this.url);
  @override
  String toString() => 'FetchFailedException: $url failed';
}

Future<String> fetchAsync(String url, int delayMs, {bool shouldFail = false}) async {
  await Future.delayed(Duration(milliseconds: delayMs)); // suspends WITHOUT blocking the event loop
  if (shouldFail) throw FetchFailedException(url);
  return '$url -> 200 OK';
}

void main() async {
  print('--- Kicking off 4 concurrent fetches ---');
  var start = DateTime.now();

  // Futures start running the MOMENT fetchAsync() is called, not when awaited --
  // calling all four before awaiting any of them is what makes them concurrent.
  // eagerError: false means Future.wait won't short-circuit on the FIRST failure;
  // it waits for every future to settle, wrapping each in a Result via .then/.catchError
  // below so one failure can't silently swallow the others' outcomes.
  var futures = [
    fetchAsync('https://api.example.com/users', 300),
    fetchAsync('https://api.example.com/orders', 500),
    fetchAsync('https://api.example.com/broken', 200, shouldFail: true),
    fetchAsync('https://api.example.com/products', 400),
  ];

  var outcomes = await Future.wait(
    futures.map(
      (f) => f.then<MapEntry<bool, String>>(
        (value) => MapEntry(true, value),
        onError: (Object e) => MapEntry(false, e.toString()),
      ),
    ),
    eagerError: false,
  );

  var elapsedMs = DateTime.now().difference(start).inMilliseconds;
  print('Elapsed: ${elapsedMs}ms (slowest single delay was 500ms -- proves concurrency,'
      ' not a ~1400ms sum)');

  print('\n--- per-fetch results ---');
  for (var outcome in outcomes) {
    var status = outcome.key ? 'OK' : 'FAILED';
    print('  $status: ${outcome.value}');
  }

  var failures = outcomes.where((o) => !o.key).length;
  print('\n${outcomes.length - failures} succeeded, $failures failed -- no result silently dropped.');
}
